File: spsc-lite-ruby/ResidualProgramGenerator.rb
require 'SLL'
require 'Algebra'
require 'ProcessTree'

class ResidualProgramGenerator
  include SLL
  include Algebra
  def initialize(tree)
    @tree = tree
    @sigs = Hash.new
    @rules = []
  end

  def genResidualProgram()
    resExp = genExp(@tree.root)
    [Program.new(@rules), resExp]
  end

  def genExp(beta)
    alpha = beta.funcAncestor()
    exp = beta.exp
    if not alpha
      if exp.isVar():
        exp
      elsif exp.isCtr()
        resExpList = genExpList(beta.children)
        Ctr.new(exp.name, resExpList)
      elsif exp.isFGCall()
        genCall(beta)
      elsif exp.isLet()
        resExpList = genExpList(beta.children)
        vnames = exp.bindings.map{|b| b[0]}
        x = vnames.zip(resExpList[1..-1])
        subst = assoc_to_h(vnames.zip(resExpList[1..-1]))
        resExpList[0].applySubst(subst)
      else
        raise "Invalid expression"
      end
    else
      name, params = @sigs[alpha]
      args = params.map{|param| Var.new(param)}
      subst = matchAgainst(alpha.exp, beta.exp)
      contr = alpha.children[0].contr
      if not contr
        return FCall.new(name, args).applySubst(subst)
      else
        return GCall.new(name, args).applySubst(subst)
      end
    end
  end

  def genExpList(nodes)
    nodes.map {|node| genExp(node) }
  end

  def isVarTest(beta)
    beta.children[0].contr ? true : false
  end

  def getFGSig(prefix, beta, name, vs)
    sig = @sigs[beta]
    if sig
      sig
    else
      name1 = "#{prefix}#{name[1..-1]}#{@sigs.length + 1}"
      sig1 = [name1, vs]
      @sigs[beta] = sig1
      sig1
    end
  end

  def getChContr(children)
    children.map{|n| [n.contr.cname, n.contr.cparams]}
  end

  def genCall(beta)
    exp = beta.exp
    name = exp.name
    args = exp.args
    params = exp.vars()
    if isVarTest(beta)
      name1, vs1 = getFGSig("g", beta, name, params)
      bodies = genExpList(beta.children)
      contrs = getChContr(beta.children)
      grules = contrs.zip(bodies).map do |contr1, body1|
        cname1, cparams1 = contr1
        GRule.new(name1, cname1, cparams1, params[1..-1], body1)
      end
      @rules += grules
      GCall.new(name1, params.map{|param| Var.new(param)})
    elsif @tree.isFuncNode(beta)
      name1, vs1 = getFGSig("f", beta, name, params)
      body1 = genExp(beta.children[0])
      @rules += FRule.new(name1, params1, body1)
      FCall.new(name1, params.map{|param| Var.new(param)})
    else
      genExp(beta.children[0])
    end
  end
end

File: spsc-lite-julia/src/residual_program_generator.jl
module ResidualProgramGenerator

using SPSC.SLanguage
using SPSC.Algebra
using SPSC.ProcessTree

const Sig = Tuple{Name,Vector{Name}}
const Sigs = Dict{NodeId,Sig}

struct ResPrGen
    tree::Tree
    sigs::Sigs
    rules::Vector{Rule}
end

ResPrGen(tree::Tree) = ResPrGen(tree, Sigs(), Rule[])

function genResidualProgram(tree::Tree)
    g = ResPrGen(tree)
    resExp = genExp!(g, tree.root)
    return (Program(g.rules), resExp)
end

function genExp!(g::ResPrGen, beta::Node)::Exp
    genExp!(g, beta, funcAncestor(beta))
end

function genExp!(g::ResPrGen, beta::Node, alpha::Node)::Exp
    (name, params) = g.sigs[alpha.nodeId]
    args = [Var(param) for param in params]
    subst = matchAgainst(alpha.e, beta.e)
    contr = alpha.children[1].contr
    if contr === nothing
        return applySubst(subst, CFG(FCall, name, args))
    else
        return applySubst(subst, CFG(GCall, name, args))
    end
end

function genExp!(g::ResPrGen, beta::Node, ::Nothing)::Exp
    genExp!Beta(g, beta, beta.e)
end

function genExp!Beta(g::ResPrGen, beta::Node, e::Var)::Exp
    return e
end

function genExp!Beta(g::ResPrGen, beta::Node, e::CFG)::Exp
    if isCtr(e)
        resExps = Exp[genExp!(g, n) for n in beta.children]
        return CFG(Ctr, e.name, resExps)
    else
        return genCall!(g, beta)
    end
end

function genExp!Beta(g::ResPrGen, beta::Node, e::Let)::Exp
    resExpList = Exp[genExp!(g, n) for n in beta.children]
    vnames = [b.name for b in e.bindings]
    subst = Dict(zip(vnames, resExpList[2:end]))
    return applySubst(subst, resExpList[1])
end

function isVarTest(beta::Node)::Bool
    beta.children[1].contr !== nothing
end

function getFGSig!(g, prefix, beta, name, vs)
    sig = get(g.sigs, beta.nodeId, nothing)
    if sig !== nothing
        return sig
    else
        name1 = "$(prefix)$(name[2:end])$(length(g.sigs) + 1)"
        sig1 = (name1, vs)
        g.sigs[beta.nodeId] = sig1
        return sig1
    end
end

function genCall!(g::ResPrGen, beta::Node)
    e = beta.e
    name = e.name
    args = e.args
    params = vars(e)
    if isVarTest(beta)
        (name1, vs1) = getFGSig!(g, "g", beta, name, params)
        bodies = [genExp!(g, n) for n in beta.children]
        contrs = [(n.contr.cname, n.contr.cparams)
                  for n in beta.children]
        grules = [GRule(name1, cname1, cparams1, params[2:end], body1)
                  for ((cname1, cparams1), body1) in zip(contrs, bodies)]
        append!(g.rules, grules)
        return CFG(GCall, name1, [Var(param) for param in params])
    elseif isFuncNode(g.tree, beta)
        (name1, params1) = getFGSig!(g, "f", beta, name, params)
        body1 = genExp!(g, beta.children[1])
        push!(g.rules, FRule(name1, params1, body1))
        return CFG(FCall, name1, [Var(param) for param in params])
    else
        return genExp!(g, beta.children[1])
    end
end

export genResidualProgram

end
